activerecord/lib/active_record/associations/builder/belongs_to.rb

```diff
@@ -71,18 +71,18 @@ def self.add_counter_cache_callbacks(model, reflection)
       klass.attr_readonly cache_column if klass && klass.respond_to?(:attr_readonly)
     end
 
-    def self.touch_record(o, changes, foreign_key, name, touch, touch_method) # :nodoc:
+    def self.touch_record(o, changes, reflection, touch_method) # :nodoc:
+      foreign_key = reflection.foreign_key
+      touch = reflection.options[:touch]
       old_foreign_id = changes[foreign_key] && changes[foreign_key].first
 
       if old_foreign_id
-        association = o.association(name)
-        reflection = association.reflection
         if reflection.polymorphic?
           foreign_type = reflection.foreign_type
           klass = changes[foreign_type] && changes[foreign_type].first || o.public_send(foreign_type)
           klass = klass.constantize
         else
-          klass = association.klass
+          klass = reflection.klass
```

Comment:
`reflection.klass` is not always the same with `association.klass`.

https://github.com/rails/rails/blob/fc7a6c738122ba9dd6eff4ede08d67e86a3f0ea7/activerecord/lib/active_record/associations/belongs_to_polymorphic_association.rb#L6-L10